diff --git a/llvm/lib/Target/Toy/ToyInstrInfo.td b/llvm/lib/Target/Toy/ToyInstrInfo.td
index e4506d9e5522..0d2470932566 100644
--- a/llvm/lib/Target/Toy/ToyInstrInfo.td
+++ b/llvm/lib/Target/Toy/ToyInstrInfo.td
@@ -164,6 +164,12 @@ def FSUBS : ArithLogicF<0b0000100, 0b000, "fsub.s", fsub>;
 def FMULS : ArithLogicF<0b0001000, 0b000, "fmul.s", fmul>;
 def FDIVS : ArithLogicF<0b0001100, 0b000, "fdiv.s", fdiv>;
 
+def FMADDS : InstR4<0b1000011, 0b00, 0b000, (outs FPR:$rd), (ins FPR:$rs1, FPR:$rs2, FPR: $rs3),
+      "fmadd.s\t$rd, $rs1, $rs2, $rs3",
+      [(set FPR:$rd, (fma FPR:$rs1, FPR:$rs2, FPR:$rs3))]>;
+
+def FMAXNUM : ArithLogicF<0b0010100, 0b001, "fmax.s", fmaxnum>;
+
 class ArithLogicF64 <bits<7> funct7, bits<3> funct3, string inst, SDNode node>:
       InstR<0b1010011, funct7, funct3, (outs FPR64:$rd), (ins FPR64:$rs1, FPR64:$rs2),
       !strconcat(inst, "\t$rd, $rs1, $rs2"),
@@ -173,3 +179,7 @@ def FADDD : ArithLogicF64<0b0000001, 0b000, "fadd.d", fadd>;
 def FSUBD : ArithLogicF64<0b0000101, 0b000, "fsub.d", fsub>;
 def FMULD : ArithLogicF64<0b0001001, 0b000, "fmul.d", fmul>;
 def FDIVD : ArithLogicF64<0b0001101, 0b000, "fdiv.d", fdiv>;
+
+def FMADDD : InstR4<0b1000011, 0b01, 0b000, (outs FPR64:$rd), (ins FPR64:$rs1, FPR64:$rs2, FPR64: $rs3),
+      "fmadd.d\t$rd, $rs1, $rs2, $rs3",
+      [(set FPR64:$rd, (fma FPR64:$rs1, FPR64:$rs2, FPR64:$rs3))]>;
